app/crud.py

```diff
@@ -0,0 +1,295 @@
+    db_pipeline.steps = jsonable_encoder(step_ids)
+    db_pipeline.status = "spawned"
+    db.commit()
+    return db_pipeline
+
+
+def get_steps_of_pipeline(db: Session, pipeline_id: int):
+    """Fetches the Steps corresponding to the Pipeline."""
+    db_pipeline = db.query(models.Pipeline).\
+        filter(models.Pipeline.id == pipeline_id).first()
+    steps = []
+    for step_id in db_pipeline.steps:
+        steps.append(get_step_by_id(db, step_id))
+    return steps
+
+
+def autorun_pipelines(db: Session):
+    """
+    Runs the next step according to the autorun logic.
+
+    - Only Pipelines with the status "autorun" are considered.
+    - Pipelines are then ordered by their ids.
+    - Once a candidate Pipeline has been selected we examine its Steps:
+    -- We skip steps which are "completed".
+    -- If we find a step already "running" then the autorunner process waits
+    for its completion. If it succeeds it is skipped. If it fails then the
+    Pipeline itself is skipped for now.
+    -- If we find a step marked as "failed" we skip this Pipeline.
+    -- If we get to a "prelaunch" status step, we execute that and wait
+    for its resolution (success or failure) and handle it accordingly.
+    -- If we are either done with the Steps of a given Pipeline or found a
+    failed one, we proceed to the next Pipeline.
+
+    If the last Step of a Pipeline is completed, then we change the Pipeline!s
+    status from "autorun" to "completed".
+    """
+    pipes = get_pipelines_by_status(db, status="autorun")
+    for pipe in pipes:
+        print(f"Attempting to progress pipeline #{pipe.id}")
+        if is_pipe_ready(db, pipe.id):
+            steps = get_steps_of_pipeline(db, pipe.id)
+            encountered_failure = False
+            for step in steps:
+                if step.status == "completed":
+                    print(f"--Step #{step.id} was already completed.")
+                    pass
+                elif step.status == "running":
+                    print(f"--Step #{step.id} was already started.")
+                    # We have to wait until it finishes
+                    while step.status == "running":
+                        print(f"--Waiting for #{step.id} to complete.")
+                        time.sleep(10)
+                        db.refresh(step)
+                    if step.status == "failed":
+                        print(f"--Step #{step.id} failed.")
+                        encountered_failure = True
+                        break
+                elif step.status == "failed":
+                    # We cannot progress with this pipeline:
+                    print(f"--Step #{step.id} failed.")
+                    encountered_failure = True
+                    break
+                else:
+                    # We have to run this step:
+                    step.run_script()
+                    step.status = "running"
+                    db.commit()

```

Comment:
Commit:
We have to commit that change in step status into the DB right there.

Because when the background execution of the script in run_script finishes, it will call back to the API and will set the step status from "running" to either "complete" or "failed".

Try-except-finally: I don't think so. Since this is a manager webapp, I think we are fine with crashing if there is a DB-related error. 
The worst thing that could happen (at least what came to my mind) is that a background task completes a step and there is no app waiting for its completed signal, and thus the fact of the completion is not noted in the DB, so we have to redo it. 
But if there is an error in the DB with updating records, then we could hardly avoid that problem with a try-except-finally.